Rubberduck.Core/UI/SelectionChangeService.cs

```diff
@@ -33,47 +34,60 @@ public SelectionChangeService(IVBE vbe, IParseCoordinator parser)
             VBENativeServices.WindowFocusChange += OnVbeFocusChanged;
         }
         
-        private void OnVbeSelectionChanged(object sender, SelectionChangedEventArgs e)
+        private void OnVbeSelectionChanged(object sender, EventArgs e)
         {
-            if (e.CodePane == null || e.CodePane.IsWrappingNullReference)
-            {
-                return;
-            }
-
             new Task(() =>
             {
-                var eventArgs = new DeclarationChangedEventArgs(e.CodePane, _parser.State.FindSelectedDeclaration(e.CodePane));
-                DispatchSelectedDeclaration(eventArgs);
+                using (var active = _vbe.ActiveCodePane)
+                {
+                    if (active == null)
+                    {
+                        return;
+                    }
+                    var eventArgs = new DeclarationChangedEventArgs(_parser.State.FindSelectedDeclaration(active));
+                    DispatchSelectedDeclaration(eventArgs);
+                }
             }).Start();
         }
 
         private void OnVbeFocusChanged(object sender, WindowChangedEventArgs e)
         {
             if (e.EventType == FocusType.GotFocus)
             {
-                switch (e.Window.Type)
+                switch (e.Hwnd.ToWindowType())
                 {
-                    case WindowKind.Designer:
-                        //Designer or control on designer form selected.
-                        if (e.Window == null || e.Window.IsWrappingNullReference || e.Window.Type != WindowKind.Designer)
+                    case WindowType.DesignerWindow:
+                        new Task(() =>
```

Comment:
I'm not sure if case 1 is possible at this point in that the only thing that the `SelectionChangeService` is driving is the toolbar. It *might* be a problem if it is eventually wired up to something that changes the selection.

As for case 2, given the current use the only implication that would have would be flicker? on the toobar as the selection switches back and forth.

Note that the reason these are spun off into tasks is because this isn't on the UI thread.

Reply:
Wait a minute; you want to run them on the UI thread? If that's the case, we should be using `UiDispatcher` here. 

As the comments - I'm thinking that the subclassing event handling needs to be handled in a general manner that is unlikely to cause problems for any potential consumers, as I expect that we will expand more and more beyond the AC, the status tool, whatever. So I'm thinking we want it to be quite robust and won't trip over its own feet even if we shoot our feet in trying to do so. 😆

Reply:
@bclothier Hmmm... that's a catch 22 then. If we want this handled in a general manner, we probably *shouldn't* use the `UiDispatcher`, no?  Regardless, if these are blocking that's a ***big*** problem due to the amount of traffic through here.

I don't disagree with the need to make this more robust though, but I think that can be addressed when we expand beyond AC and the toolbar. It won't be our feet that we shoot - it will be much higher...

Reply:
Agreed that we don't want it on UI thread at all; running the task in their own thread is probably best overall, but because I can see events firing so frequently and so rapidly, we will see weird things if we don't manage that and for that reason, one task per event raised might not be the optimal strategy. Consider also that `Task` are known to re-use threads which may have its own ramifications as well. #BeltsAndSuspenders